### ISSUE TITLE
Enabling a baseline hides output when there are passing checks only

### ISSUE THREAD

**Issue by ferrarimarco** [bridgecrewio/checkov#5873]
**Describe the issue**

When I set baseline, Checkov doesn't report any output if there are only passing checks when running against a directory or a file. The contents of the `failed_checks` array in the baseline file don't matter. This happens even if the `failed_checks` array is empty.

Here is a reproduction case.

`.checkov.baseline` file:

```json
{
    "failed_checks": []
}
```

`example.tf` file:

```hcl
resource "aws_instance" "good" {
  ami                         = "ami-0ff8a91507f77f867"
  associate_public_ip_address = false
  ebs_optimized               = true
  iam_instance_profile        = "test"
  instance_type               = "t2.small"
  monitoring                  = true
  vpc_security_group_ids      = ["sg-12345678901234567"]

  metadata_options {
    http_endpoint = "disabled"
  }

  ebs_block_device {
    device_name = "name"
    encrypted   = true
  }

  root_block_device {
    encrypted = true
  }
}

terraform {
  required_version = ">=1.0"
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 4.0"
    }
  }
}

```

Run Checkov `3.1.38` with the following command:

```
checkov --baseline /path/to/.checkov.baseline --file /path/to/example.tf
```

Output:

```sh
[ terraform framework ]: 100%|████████████████████|[1/1], Current File Scanned=test/linters/checkov/good/terraform_checkov_good_1.tf
[ secrets framework ]: 100%|████████████████████|[1/1], Current File Scanned=/tmp/lint/test/linters/checkov/good/terraform_checkov_good_1.tf

       _               _
   ___| |__   ___  ___| | _______   __
  / __| '_ \ / _ \/ __| |/ / _ \ \ / /
 | (__| | | |  __/ (__|   < (_) \ V /
  \___|_| |_|\___|\___|_|\_\___/ \_/

By Prisma Cloud | version: 3.1.38
```

```sh
checkov --file /path/to/example.tf
```

Output:

```
[ terraform framework ]: 100%|████████████████████|[1/1], Current File Scanned=test/linters/checkov/good/terraform_checkov_good_1.tf
[ secrets framework ]: 100%|████████████████████|[1/1], Current File Scanned=/tmp/lint/test/linters/checkov/good/terraform_checkov_good_1.tf

       _               _
   ___| |__   ___  ___| | _______   __
  / __| '_ \ / _ \/ __| |/ / _ \ \ / /
 | (__| | | |  __/ (__|   < (_) \ V /
  \___|_| |_|\___|\___|_|\_\___/ \_/

By Prisma Cloud | version: 3.1.38

terraform scan results:

Passed checks: 7, Failed checks: 0, Skipped checks: 0

Check: CKV_AWS_126: "Ensure that detailed monitoring is enabled for EC2 instances"
	PASSED for resource: aws_instance.good
	File: /terraform_checkov_good_1.tf:1-22
	Guide: https://docs.prismacloud.io/en/enterprise-edition/policy-reference/aws-policies/aws-logging-policies/ensure-that-detailed-monitoring-is-enabled-for-ec2-instances
Check: CKV_AWS_88: "EC2 instance should not have public IP."
	PASSED for resource: aws_instance.good
	File: /terraform_checkov_good_1.tf:1-22
	Guide: https://docs.prismacloud.io/en/enterprise-edition/policy-reference/aws-policies/public-policies/public-12
Check: CKV_AWS_46: "Ensure no hard-coded secrets exist in EC2 user data"
	PASSED for resource: aws_instance.good
	File: /terraform_checkov_good_1.tf:1-22
	Guide: https://docs.prismacloud.io/en/enterprise-edition/policy-reference/aws-policies/secrets-policies/bc-aws-secrets-1
Check: CKV_AWS_79: "Ensure Instance Metadata Service Version 1 is not enabled"
	PASSED for resource: aws_instance.good
	File: /terraform_checkov_good_1.tf:1-22
	Guide: https://docs.prismacloud.io/en/enterprise-edition/policy-reference/aws-policies/aws-general-policies/bc-aws-general-31
Check: CKV_AWS_135: "Ensure that EC2 is EBS optimized"
	PASSED for resource: aws_instance.good
	File: /terraform_checkov_good_1.tf:1-22
	Guide: https://docs.prismacloud.io/en/enterprise-edition/policy-reference/aws-policies/aws-general-policies/ensure-that-ec2-is-ebs-optimized
Check: CKV_AWS_8: "Ensure all data stored in the Launch configuration or instance Elastic Blocks Store is securely encrypted"
	PASSED for resource: aws_instance.good
	File: /terraform_checkov_good_1.tf:1-22
	Guide: https://docs.prismacloud.io/en/enterprise-edition/policy-reference/aws-policies/aws-general-policies/general-13
Check: CKV2_AWS_41: "Ensure an IAM role is attached to EC2 instance"
	PASSED for resource: aws_instance.good
	File: /terraform_checkov_good_1.tf:1-22
	Guide: https://docs.prismacloud.io/en/enterprise-edition/policy-reference/aws-policies/aws-iam-policies/ensure-an-iam-role-is-attached-to-ec2-instance
```

**Additional context**

N/A

Thanks for your support!

**Comment by gruebel** [comment]
hey @ferrarimarco thanks for reaching out.

I'm not 100% sure, but it looks intentional `checkov` not showing passing checks for checks, which are not in the baseline, but I can understand this being a bit surprising. It is quite easy to fix, but it seems to be a "breaking change", if we adjust the current behaviour.

We also add an info at the bottom of the run, which kind of indicates this behaviour

> Baseline analysis report using .checkov.baseline - only new failed checks with respect to the baseline are reported

@tsmithv11 @Saarett any thoughts about it?

**Comment by ferrarimarco** [comment]
Hi @gruebel !

This is somewhat surprising also considering that it happens even when the `failed_checks` array is empty.

Thanks for following up!

**Comment by tsmithv11** [comment]
@gruebel @ferrarimarco I'm told by the engineer who implemented the baseline feature that we did not include passing checks to save space.

**Comment by krizi** [comment]
would it be possible to verify this again? would be great if passed checks would still appear in the output (eg. junit) even if a baseline is provided. would help to provide an evidence which tests has been checked against the sources.

**Comment by Saarett** [comment]
Hi @krizi , we have no plans to change this.

**Comment by tomasz-baran** [comment]
Hi @gruebel, @tsmithv11 and @Saarett.

I have an impression that Checkov is already awesome for Terraform resources but not for external modules YET 😉

My use case is that I would like to switch from a few small Terraform resources for an S3 bucket (aws_s3_bucket, aws_s3_bucket_ownership_controls, aws_s3_bucket_public_access_block) to a single module (https://registry.terraform.io/modules/terraform-aws-modules/s3-bucket/aws/latest) but keeping some polices disabled (we disabled them by inline comments, e.g. #checkov:skip=CKV_AWS_21:versioning not required).

The thing is that the S3 module is generic and external (public) so I cannot add such inline comments there. The only option I found is to scan the Terraform plan file instead and disable too strict policies (too strict just in our case) with the baseline file feature.

And what I miss?

1) I expected a consistent behavior i.e. having failed checks only in the output only when the `--quiet` option is provided.
2) No information about the real number of tests, when I have a plan file that produces just a single failed test, it looks like checkov reports only that singe test, even with `-o junitxml` (`<testsuite disabled="0" errors="0" failures="1" name="terraform_plan scan" skipped="0" tests="1" time="0">`). And why skipped=0 since I have multiple `check_ids` defined in the baseline file.

Or maybe It would be better to add a functionality that would allow to ignore checks for modules? Similar to the baseline file but defined in YAML that would allow to have comments why a policy is disabled (json used for the baseline file doesn't accept any comments). And the disabled checks that way could be reported as "skipped" in the junitxml output.

Or do you have any suggestions how I can handle my case in a better way? 😊

**Comment by tsmithv11** [comment]
Hey @tomasz-baran thanks for reaching out. I'll give a partial answer focused on skipping a check. We have a few options - you can either add an inline skip to the module block and then run checkov using the `--download-external-modules true` flag. For plan files, you can't add inline comments, so you can use the `--repo-root-for-plan-enrichment` and add the skip comment to the module block. This will skip the check for the resources in the module and not globally.

**Comment by tomasz-baran** [comment]
Thank @tsmithv11 for the helpful reply! I didn't know that inline skips are supported also inside module blocks. There is no such example on the `Suppressing and Skipping Policies` docs page.

This method works well with a terraform plan since it gives me the same results as when using just terraform resources.

One thing that it would be nice to improve is to not display the following warning when the `--download-external-modules false` option is used:

```
checkov --quiet --compact --framework terraform_plan --repo-root-for-plan-enrichment .  -f tfplan.json --download-external-modules false
2024-11-25 20:19:52,815 [MainThread  ] [WARNI]  Failed to download module terraform-aws-modules/s3-bucket/aws:~> 4.2.2 (for external modules, the --download-external-modules flag is required)
terraform_plan scan results:

Passed checks: 11, Failed checks: 0, Skipped checks: 6
```

Here is the code I use for testing and comparison:

**1) Case 1: Resources**

```
resource "aws_s3_bucket" "example1" {
  bucket = "tb-tf-resources"

  #checkov:skip=CKV_AWS_18:Audit logging can be enabled in the future
  #checkov:skip=CKV_AWS_21:Object versioning not required
  #checkov:skip=CKV2_AWS_61:Lifecycle configuration not required
  #checkov:skip=CKV2_AWS_62:Event notifications not required
  #checkov:skip=CKV_AWS_144:S3 Replication not required
  #checkov:skip=CKV_AWS_145:TODO: Switch to KMS
}

resource "aws_s3_bucket_ownership_controls" "example1" {
  bucket = aws_s3_bucket.example1.bucket

  rule {
    object_ownership = "BucketOwnerEnforced"
  }
}

resource "aws_s3_bucket_public_access_block" "example1" {
  bucket = aws_s3_bucket.example1.bucket

  block_public_acls       = true
  block_public_policy     = true
  ignore_public_acls      = true
  restrict_public_buckets = true
}
```

Results:

```
checkov --quiet --compact --framework terraform -d .
terraform scan results:

Passed checks: 11, Failed checks: 0, Skipped checks: 6
```

**2) Case 2: External module**

```
module "s3_bucket" {
  source  = "terraform-aws-modules/s3-bucket/aws"
  version = "~> 4.2.2"

  bucket = "tb-tf-module"

  control_object_ownership = true
  object_ownership         = "BucketOwnerEnforced"

  # Block Public Access
  block_public_acls       = true
  block_public_policy     = true
  ignore_public_acls      = true
  restrict_public_buckets = true

  #checkov:skip=CKV_TF_1:using a module from the Terraform registry where a commit hash is not applicable

  #checkov:skip=CKV_AWS_18:Audit logging can be enabled in the future
  #checkov:skip=CKV_AWS_21:Object versioning not required
  #checkov:skip=CKV2_AWS_61:Lifecycle configuration not required
  #checkov:skip=CKV2_AWS_62:Event notifications not required
  #checkov:skip=CKV_AWS_144:S3 Replication not required
  #checkov:skip=CKV_AWS_145:TODO: Switch to KMS
}
```

Results:

```
Passed checks: 43, Failed checks: 2, Skipped checks: 7

Check: CKV_AWS_300: "Ensure S3 lifecycle configuration sets period for aborting failed uploads"
        FAILED for resource: module.s3_bucket.aws_s3_bucket_lifecycle_configuration.this[0]
        File: /.external_modules/github.com/terraform-aws-modules/terraform-aws-s3-bucket/8b855f886e3f7f27ea4cdb692c94805fdf25f9e3/main.tf:240-354
        Calling File: /main.tf:1-24
        Guide: https://docs.prismacloud.io/en/enterprise-edition/policy-reference/aws-policies/aws-general-policies/bc-aws-300
Check: CKV2_AWS_6: "Ensure that S3 bucket has a Public Access block"
        FAILED for resource: module.s3_bucket.aws_s3_bucket.this[0]
        File: /.external_modules/github.com/terraform-aws-modules/terraform-aws-s3-bucket/8b855f886e3f7f27ea4cdb692c94805fdf25f9e3/main.tf:25-34
        Guide: https://docs.prismacloud.io/en/enterprise-edition/policy-reference/aws-policies/aws-networking-policies/s3-bucket-should-have-public-access-blocks-defaults-to-false-if-the-public-access-block-is-not-attached
```

So depending on the method I get different results. I think it depends on how a module is written but in this case Checkov reports CKV2_AWS_6 even though the public access is disabled (via [aws_s3_bucket_public_access_block](https://github.com/terraform-aws-modules/terraform-aws-s3-bucket/blob/8b855f886e3f7f27ea4cdb692c94805fdf25f9e3/main.tf#L913) under the hood). For that reason, I'm going to use the plan based scan, at least for now.
